lib/chef/knife/ec2_base.rb

```diff
@@ -327,7 +327,7 @@ def validate_aws_credential_file!
 
       aws_creds = ini_parse(File.read(aws_cred_file_location))
       profile = locate_config_value(:aws_profile)
-
+      profile = "profile #{profile}" if profile != "default"
```

Comment:
may be we can write like this to avoid two time variable assignment 
profile = "profile #{locate_config_value(:aws_profile)}" unless locate_config_value(:aws_profile) == "default"

Reply:
@Vasu1105 Thank you, I do agree with your point that reassignment should be avoided but here it is not the case of reassignment.

We are appending a specific tag in existing string based on a certain condition.
So assigning a variable based on the condition returns `nil` if condition is not met. 
Also calling un memoize method twice is not recommended.

Reply:
@dheerajd-msys Thank you. Make sense.

Reply:
Tested this patch, it looks like it's working for me.

Reply:
ok, Thank you @breisig for confirmation.

Reply:
Actually, I spoke too soon. I get another error now when it's trying to bootstrap.

```
root@server ~# knife ec2 server create 10.50.14.140 -N server1.mycompany.com --image ami-0b41608dde0cdd48a -f m5.large --region ca-central-1 --ssh-key myssh-ssh-aws --subnet subnet-d458fab4 -g sg-b4d0cdc2 -E production -r 'role[baseinstall]' --ebs-optimized --ebs-size 100 -A --server-connect-attribute private_ip_address --private-ip-address 10.50.14.140 --json-attributes='{"tags": ["app:myapp"]}' --aws-profile mycompany --node-ssl-verify-mode none --ssh-verify-host-key never -U root

Waiting for EC2 to create the instance
Instance ID: i-693acb4f5557a43b2
Flavor: m5.large
Image: ami-0b41608dde0cdd48a
Region: ca-central-1
Availability Zone: ca-central-1a
Security Group Ids: sg-b4d0cdc2
AWS Tags: Name: server1.mycompany.com
SSH Key: myssh-ssh-aws
T2/T3 Unlimited: Disabled

Subnet ID: subnet-d158ffb8
Tenancy: default
Private IP Address: 10.50.14.140

Waiting for sshd access to become available
SSH Target Address: 10.50.14.140(private_ip_address)
done

SSH Target Address: 10.50.14.140(private_ip_address)
Connecting to
ERROR: Train::ClientError: You must provide a value for "host".
```
Not sure what this 'ERROR: Train::ClientError: You must provide a value for "host".' is all about now.

Reply:
Hi @breisig From your command it looks like it is missing `--associate-public-ip` flag.
Please try to run your command adding this attribute and let us know if there is an issue.

Thank you!

Reply:
@dheerajd-msys I tried adding --associate-public-ip and it looks better but it's not connecting to the right host. That command makes chef create a public IP even though the IP of the server is an internal private lan ip [10.50.14.140]. It should be trying to connect to 10.50.14.140 as in the '10.50.14.140 -N server1.mycompany.com' and '--server-connect-attribute private_ip_address --private-ip-address 10.50.14.140'. That should make chef connect to 10.50.14.140 to bootstrap that server.

```
root@server ~# knife ec2 server create 10.50.14.140 -N server1.mycompany.com --image ami-0b41608dde0cdd48a -f m5.large --region ca-central-1 --ssh-key myssh-ssh-aws --subnet subnet-d458fab4 -g sg-b4d0cdc2 -E production -r 'role[baseinstall]' --ebs-optimized --ebs-size 100 -A --server-connect-attribute private_ip_address --private-ip-address 10.50.14.140 --json-attributes='{"tags": ["app:myapp"]}' --aws-profile mycompany --node-ssl-verify-mode none --ssh-verify-host-key never -U root  --associate-public-ip

Waiting for EC2 to create the instance
Instance ID: i-693acb4f5557a43b2
Flavor: m5.large
Image: ami-0b41608dde0cdd48a
Region: ca-central-1
Availability Zone: ca-central-1a
Security Group Ids: sg-b4d0cdc2
AWS Tags: Name: server1.mycompany.com
SSH Key: myssh-ssh-aws
T2/T3 Unlimited: Disabled

Subnet ID: subnet-d458fab4
Tenancy: default
Public DNS Name: ec2-65-123-104-521.ca-central-1.compute.amazonaws.com
Private IP Address: 10.50.14.140

Waiting for sshd access to become available
SSH Target Address: 10.50.14.140(private_ip_address)
done

SSH Target Address: 10.50.14.140(private_ip_address)
Connecting to ec2-65-123-104-521.ca-central-1.compute.amazonaws.com
WARN: [SSH] connection failed, terminating (#<Net::SSH::ConnectionTimeout: Net::SSH::ConnectionTimeout>)
ERROR: Train::Transports::SSHFailed: SSH session could not be established
```

Reply:
@breisig Please try with user other than `root` and include `--sudo` in command. 
There are specific username defined as per AWS AMI e.g. `ec2-user` for Amazon Linux AMI and `ubuntu` for Ubuntu AMI.

Reference : https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/AccessingInstancesLinux.html